src/boot/measure.c

```diff
@@ -464,6 +525,235 @@ static int verb_calculate(int argc, char *argv[], void *userdata) {
+        if (!pubkey)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to parse public key '%s'.", arg_public_key);
+
+        r = pcr_states_allocate(&pcr_states);
+        if (r < 0)
+                return r;
+
+        n = (size_t) r;
+
+        r = measure_pcr(pcr_states, n);
+        if (r < 0)
+                return r;
+
+        r = dlopen_tpm2();
+        if (r < 0)
+                return r;
+
+        r = tpm2_context_init(arg_tpm2_device, &c);
+        if (r < 0)
+                return r;
+
+        for (size_t i = 0; i < n; i++) {
+                static const TPMT_SYM_DEF symmetric = {
+                        .algorithm = TPM2_ALG_AES,
+                        .keyBits.aes = 128,
+                        .mode.aes = TPM2_ALG_CFB,
+                };
+                PcrState *p = pcr_states + i;
+
+                rc = sym_Esys_StartAuthSession(
+                                c.esys_context,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                NULL,
+                                TPM2_SE_TRIAL,
+                                &symmetric,
+                                TPM2_ALG_SHA256,
+                                &session_handle);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to open session in TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
+
+                /* Generate a single hash value from the PCRs included in our policy. Given that that's
+                 * exactly one, the calculation is trivial. */
+                TPM2B_DIGEST intermediate_digest = {
+                        .size = SHA256_DIGEST_SIZE,
+                };
+                assert(sizeof(intermediate_digest.buffer) >= SHA256_DIGEST_SIZE);
+                sha256_direct(p->value, p->value_size, intermediate_digest.buffer);
+
+                TPM2_ALG_ID tpmalg = convert_evp_md_name_to_tpm2_alg(p->md);
+                if (tpmalg == TPM2_ALG_ERROR) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Unsupported PCR bank");
+                        goto finish;
+                }
+
+                TPML_PCR_SELECTION pcr_selection;
+                tpm2_pcr_mask_to_selection(1 << TPM_PCR_INDEX_KERNEL_IMAGE, tpmalg, &pcr_selection);
+
+                rc = sym_Esys_PolicyPCR(
+                                c.esys_context,
+                                session_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &intermediate_digest,
+                                &pcr_selection);
```

Comment:
AFAICT `measure_pcr` and pcr_selection sent to `Esys_PolicyPCR` will be measured from the current running system, is that right? If so why? When I have done this I supply both the PCR selection and the PCR values to be expected, ie calculated from a known good golden machine.

Typically I do this somewhere and have 0 dependencies from anything on the TPM. But I am likely missing something here, as usual.

Reply:
No. measure_pcr() is the function that calculates the expected PCR values from the components of a unified kernel image. The input of that function is a bunch of data files and it outputs a set of hashes for PCR 11, one for each selected bank.

I guess calculate_expected_pcr() might be a better name, if you want...

Anyway, measure_pcr() will generate data without talking to any TPM and independent of any system state. It operates solely on the kernel files you give it.

Reply:
PCR 11 because that's the PCR I grabbed for measuring unified kernel images too. it apparently isn't used on Linux systems otherwise so far. By picking a PCR nothing else uses we can pre-calculate expected values easily.

Reply:
> No. measure_pcr() is the function that calculates the expected PCR values from the components of a unified kernel image. The input of that function is a bunch of data files and it outputs a set of hashes for PCR 11, one for each selected bank.
> 
> I guess calculate_expected_pcr() might be a better name, if you want...
> 
> Anyway, measure_pcr() will generate data without talking to any TPM and independent of any system state. It operates solely on the kernel files you give it.

Ahh I see it's controlled by an argument for current:
```C
if (arg_current) {
 // do it through sysfs
else
  // do it by hand
```